### ISSUE TITLE
Deleting a SYNC replica gets MAIN stuck in kubernetes

### ISSUE THREAD

**Issue by hpdobrica** [memgraph/memgraph#1527]
**Memgraph version**
2.12.0

**Environment**
Local KIND kubernetes cluster on M1 macbook, running a slightly modified (to enable it to run multiple nodes) version of official helm chart (https://github.com/memgraph/helm-charts/pull/12)

**Describe the bug**
Creating a fresh cluster with 1 MAIN and 2 SYNC REPLICAs, and then restarting one of the REPLICAs gets the MAIN stuck.

When checking the `show replicas` on MAIN at the time of restarting a REPLICA, the REPLICA state goes from ready to invalid, shortly after which `show replicas` stops working on MAIN until MAIN is restarted. When MAIN is restarted, it actually picks up the new IP of REPLICA and continues working as expected (i.e. fully recovers).

**To Reproduce**
 
values file:
```
image:
  tag: 2.12.0

memgraphConfig:
  - "--also-log-to-stderr=true"
  - "--replication-restore-state-on-startup=true"

replicaCount: 3
```

in the helm chart from PR linked above run this:
`helm upgrade --install memgraph . -n memgraph -f values-file-from-above.yaml`

this creates 3 pods, `memgraph-0`, `memgraph-1`, `memgraph-2` - 0 will be MAIN, 1 and 2 will be replicas. Lets set that up:
```
# on both replicas run:
SET REPLICATION ROLE TO REPLICA WITH PORT 10000;

# on main run:
REGISTER REPLICA one SYNC TO "memgraph-1.memgraph.memgraph.svc.cluster.local:10000";
REGISTER REPLICA two SYNC TO "memgraph-2.memgraph.memgraph.svc.cluster.local:10000";
```

delete the pod memgraph-2. this causes the problem described above

**Expected behavior**
After pod memgraph-2 comes back, DNS resolution should be repeated e.g. with backoff:
- first time we try resolving that DNS record, it will likely fail because old pod is gone, and new one still doesn't exist
- a couple of seconds later the DNS resolution should yield a new ip and cluster should continue working as expected
right now this happens only after MAIN is restarted too

**Logs**
main logs show the following after delete of replica:
```
[2023-11-17 14:53:53.875] [memgraph_log] [error] Couldn't replicate data to two. For more details, visit https://memgr.ph/replication.
[2023-11-17 14:56:05.145] [memgraph_log] [error] [client.hpp:154] Couldn't connect to remote address 10.244.0.96:10000
[2023-11-17 14:56:05.145] [memgraph_log] [error] Failed to connect to replica two at the endpoint 10.244.0.96:10000. For more details, visit https://memgr.ph/replication.
```

**Additional context**

This ticket originated from this discord thread, more details can be found there: https://discord.com/channels/842007348272169002/890968055339958292/1175071185793450064

**Verification Environment**
Docker image, probably ARM

**Comment by antejavor** [comment]
@hpdobrica  Thanks a lot for the contributions to Helm Charts and thanks a lot for opening the issue! 🙏 🥇 

@DavIvek Looks like DNS resolution works at the beginning but it does not work if the replica is restarted.

**Comment by DavIvek** [comment]
I think that I know where the problem is. I'll check it.

**Comment by DavIvek** [comment]
It appears that we may be resolving the DNS name into an IP address prematurely. Currently, we utilize the Endpoint object to store the address and port, resolving the hostname into an IP address during the creation of the Endpoint. The issue arises when we construct an Endpoint during replica registration. Consequently, upon replica restart, it continues to use the original Endpoint, including the IP address assigned during registration. This leads to unintended error as the replica retains the initial IP address even after restarting.
A possible fix could be to change the Endpoint constructor for DNS name.
This code is in endpoint.cpp.
Changing this:
```
Endpoint::Endpoint(needs_resolving_t, std::string hostname, uint16_t port) : port(port) {
  address = ResolveHostnameIntoIpAddress(hostname, port);
  IpFamily ip_family = GetIpFamily(address);
  if (ip_family == IpFamily::NONE) {
    throw NetworkError("Not a valid IPv4 or IPv6 address: {}", address);
  }
  family = ip_family;
}
```
into: 
```
Endpoint::Endpoint(needs_resolving_t, std::string hostname, uint16_t port) : port(port) {
  address = hostname;
  IpFamily ip_family = GetIpFamily(ResolveHostnameIntoIpAddress(hostname, port));
  if (ip_family == IpFamily::NONE) {
    throw NetworkError("Not a valid IPv4 or IPv6 address: {}", address);
  }
  family = ip_family;
}
```
Also, maybe we should remove the family attribute of the Endpoint since we are not using it.

**Comment by DavIvek** [comment]
This is fixed as a part of a #1615

**Comment by karmenrabar** [comment]
Hi @hpdobrica, I'm happy to update you that Memgraph has a new release, 2.14., which includes the fix for this issue. We post all of our release updates on Discord, here's the [link](https://discord.com/channels/842007348272169002/1047507324555767858/1200107892968271956). :)
Could you check if the newest version works as expected for you now?
